### PR TITLE
feat: add smartcontract to swap before payment with any to erc20 conversion

### PR DESCRIPTION
**Pull request by vrolland**

### CODE REVIEW

File: packages/smart-contracts/test/contracts/ERC20SwapToPayWithConversion.js [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+      feeAddress: builder,
+    });
+
+    const finalBuilderBalance = await paymentNetworkErc20.balanceOf(builder);
+    const finalIssuerBalance = await paymentNetworkErc20.balanceOf(to);
+    expect(finalBuilderBalance.toNumber()).to.equals(0);
+    expect(finalIssuerBalance.toNumber()).to.equals(0);
+  });
+
+  it('cannot swap if too few payment tokens', async function () {
```
Comment (yomarion) [inline packages/smart-contracts/test/contracts/ERC20SwapToPayWithConversion.js:189]:
```suggestion
  it('cannot swap with a too low maximum spent', async function () {
```

File: packages/smart-contracts/test/contracts/ERC20SwapToPayWithConversion.js [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        builder,
+        exchangeRateOrigin + 100, // _uniswapDeadline
+        0, // _chainlinkMaxRateTimespan
+        { from },
+      ),
+    );
+    await expectPayerBalanceUnchanged();
+  });
+
+  it('cannot swap more tokens than balance', async function () {
```
[the remark this answers is not in the document]
Comment (yomarion) [reply packages/smart-contracts/test/contracts/ERC20SwapToPayWithConversion.js:249]: But could be relevant to test with a low balance, actually.

**Review by yomarion — APPROVED** [review]
You nailed that contract super fast, with proper testing, and so on, impressive ! Super good job overall.

I am not comfortable with the "direction" of both paths, I think it's odd that one goes from payment currency to network currency and the other one from request currency to network currency. But I'm not 100% sure and probably not important.

File: packages/smart-contracts/src/contracts/ERC20SwapToPayWithConversion.sol [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    uint256 _requestFeeAmount, // requestCurrency
+    address _feeAddress,
+    uint256 _uniswapDeadline,
+    uint256 _chainlinkMaxRateTimespan
+  )
+    external
+  {
+    require(
+      _uniswapPath[_uniswapPath.length-1] == _chainlinkPath[_chainlinkPath.length-1],
+      "the requested token on uniswap must be the payment currency");
```
[the remark this answers is not in the document]
Comment (yomarion) [reply packages/smart-contracts/src/contracts/ERC20SwapToPayWithConversion.sol:76]: Precising my thought, we could have:
- `[paymentToken, (some other intermediate tokens)]`
- `[requestCurrency, (some other intermediate currencies), paymentNetworkCurrency]`

From there we know that the swap path is:
`[paymentToken, (some other intermediate tokens), paymentNetworkCurrency]`

We make the developer UX a bit odd BUT we save gas by:
- Removing some input data weight
- Removing one check

Maybe the best is to offer both options? An `optimizedConvertAndSwap` that is encapsulated in a `friendlyConvertAndSwap` that performs this check ?

Or maybe keep this for later

File: packages/smart-contracts/src/contracts/ERC20SwapToPayWithConversion.sol [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+  constructor(address _swapRouterAddress, address _paymentProxyAddress) public {
+    swapRouter = IUniswapV2Router02(_swapRouterAddress);
+    paymentProxy = Erc20ConversionProxy(_paymentProxyAddress);
+    chainlinkConversionPath = ChainlinkConversionPath(paymentProxy.chainlinkConversionPath());
+  }
+
+  /**
+  * @notice Performs a token swap between a payment currency and a payment network currency after
+  *         a conversion from a request currency and then calls a payment proxy to pay
```
Comment (benjlevesque) [inline packages/smart-contracts/src/contracts/ERC20SwapToPayWithConversion.sol:42]: this is pretty complex. I honestly don't understand what it says

If it's too hard to make it clearer, give an example.

**Review by benjlevesque — APPROVED** [review]

File: packages/smart-contracts/src/contracts/ERC20SwapToPayWithConversion.sol (hunk 4 above)
Comment (benjlevesque) [reply packages/smart-contracts/src/contracts/ERC20SwapToPayWithConversion.sol:42]: Proposal:

```
Performs a request payment, denominated in a token or currency A, where the issuer expects a token B, and the payer uses a token C.
The conversion rate from A to B is done using Chainlink. 
The token swap is done using Uniswap
```

File: packages/smart-contracts/test/contracts/ERC20SwapToPayWithConversion.js [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+const TestERC20 = artifacts.require('./TestERC20.sol');
+
+const ERC20FeeProxy = artifacts.require('./ERC20FeeProxy.sol');
+
+const ChainlinkConversionPath = artifacts.require('./ChainlinkConversionPath.sol');
+const AggTest = artifacts.require('AggTest.sol');
+const Erc20ConversionProxy = artifacts.require('./Erc20ConversionProxy.sol');
+
+const FakeSwapRouter = artifacts.require('./FakeSwapRouter.sol');
```
[the remark this answers is not in the document]
Comment (yomarion) [reply packages/smart-contracts/test/contracts/ERC20SwapToPayWithConversion.js:16]: My bad, a Visual Code latency led me to think it was missing, sorry !

**Review by yomarion — APPROVED** [review]
Perfect

**Pull request merged**